WIP integrating Learning in evolution

portaloffreedom opened the pull request.

etiennegalea committed:
Simplified Deserialisation method.

etiennegalea committed:
Merge remote-tracking branch 'origin/development' into development

etiennegalea committed:
Created new protobuf message

Commit:
Skeleton for new Learner architecture

Commit:
All classes for learner abstraction in place now

Commit:
Added hardcoded parameters to BayesianOptimizer.

Commit:
Merge remote-tracking branch 'upstream/development' into development

Commit:
Refactoring

Commit:
Plugging in NoLearner

etiennegalea committed:
Added new robot_states.proto for learning evaluations

Added compiled proto files (c++ and python)

etiennegalea committed:
Merge remote-tracking branch 'origin/development' into development

Commit:
devectorize_cpg_controller and refactor

Commit:
implement optimization step and fitness saving.

Commit:
Plug new learner into robot controller

etiennegalea committed:
Merge branch 'bayes_opt' into evaluator_proto_msg

portaloffreedom committed:
Bayesian Optimization integration

portaloffreedom committed:
fix limbo compilation issue

portaloffreedom committed:
added Hyperneat learner - WIP

etiennegalea committed:
Merge branch 'learning' into evaluator_proto_msg

portaloffreedom committed:
Evaluation Reporter WIP

etiennegalea committed:
Merge remote-tracking branch 'upstream/learning' into evaluator_proto_msg

portaloffreedom committed:
node pubblisher for gazebo reporter

etiennegalea committed:
Merge remote-tracking branch 'upstream/learning' into evaluator_proto_msg

portaloffreedom committed:
improved hyperneat

and added reporter in hyperneat

etiennegalea committed:
Added protobuf construction and sending to GazeboReporter. Changed float to double in protobuf robotstateslearning.proto.

portaloffreedom committed:
Merge remote-tracking branch 'etienne/evaluator_proto_msg' into learning

etiennegalea committed:
Changed protobuf msg ID to string

portaloffreedom committed:
Merge remote-tracking branch 'etienne/evaluator_proto_msg' into learning

portaloffreedom committed:
GazeboReporter implementation ready

portaloffreedom committed:
improve learning classes

portaloffreedom committed:
removed old pygazebo and improved .gitignore

portaloffreedom committed:
WIP: new CPG BO loading integration

portaloffreedom committed:
Merge branch 'learning' of github.com:ci-group/revolve into learning

etiennegalea committed:
Fixed brain imports as it lower version of python results in errors.

Commit:
Debugging

Commit:
Merge branch 'learning' of github.com:etiennegalea/revolve into learning

portaloffreedom committed:
Fixed learning

portaloffreedom committed:
implemented brain

- C++14
- removed old DifferentialCPG
- reset robot position after evaluation
- load best controller at the end of learning
- better separated learner and controller when loading
- fixed implicit conversion of negative integer into unsigned integer

portaloffreedom committed:
new plugin that supports reports is a work in progress...

portaloffreedom committed:
Implemented new world and robot manager for the learning steps

portaloffreedom committed:
Fixed PIGPIO compile warning

portaloffreedom committed:
removed old code

portaloffreedom committed:
Fixed MultiNEAT Neural Network bug

the NN needs to be activated as many times as the network is deep

portaloffreedom committed:
Targeted locomotion for robots (HW first)

- Added AngleToTargetDetectorClass
- Added camera to HW controller
- added OpenCV dependency
- added raspicam dependency

Commit:
catch up with local

DaanZ commented on cpprevolve/revolve/brains/controller/DifferentialCPG.cpp:
```diff
@@ -52,20 +49,23 @@ using namespace revolve;
-        : next_state(nullptr)
+        const DifferentialCPG::ControllerParams &params,
+        const std::vector<std::shared_ptr<Actuator>> &actuators,
+        std::shared_ptr<AngleToTargetDetector> angle_to_target_sensor)
```
Does every future version need to specify these parameters?

DaanZ commented on cpprevolve/revolve/brains/controller/DifferentialCPG.cpp:
```diff
@@ -82,23 +82,27 @@ DifferentialCPG::DifferentialCPG(
-        const NEAT::Genome &gen)
-        : next_state(nullptr)
+        const NEAT::Genome &gen,
+        std::shared_ptr<AngleToTargetDetector> angle_to_target_sensor)
```
Is this necessary for future experiments?

DaanZ commented on cpprevolve/revolve/brains/controller/DifferentialCPG.cpp:
```diff
@@ -531,30 +573,30 @@ void DifferentialCPG::step(
+
+double DifferentialCPG::output_function(double input) const
+{
+    return this->output_signal_factor
```
Split up formula for better readability.

DaanZ commented on cpprevolve/revolve/gazebo/CMakeLists.txt:
```diff
@@ -78,7 +78,7 @@ if (LOCAL_GAZEBO_DIR)
     message(WARNING "Using local Gazebo @ ${gazebo_DIR}")
 else()
-    find_package(gazebo 9 REQUIRED)
+    find_package(gazebo 10 REQUIRED)
```
Is this change generally needed?

DaanZ commented on cpprevolve/revolve/brains/controller/DifferentialCPG.h:
```diff
@@ -40,17 +39,19 @@ class DifferentialCPG
-            const std::vector<std::shared_ptr<Actuator>> &_actuators);
+            const DifferentialCPG::ControllerParams &params,
+            const std::vector<std::shared_ptr<Actuator>> &_actuators,
+            std::shared_ptr<AngleToTargetDetector> angle_to_target_sensor = nullptr);
```
Is this needed for future experiments?

DaanZ commented on cpprevolve/revolve/gazebo/brains/Evaluator.cpp:
```diff
@@ -35,36 +38,50 @@ double Evaluator::measure_distance(
   this->previous_position_.Reset();
   this->start_position_.Reset();
-  this->locomotion_type = "directed"; // {directed, gait}
+  this->locomotion_type = "turing_left"; // {directed, gait}
```
Is this the new standard approach?

DaanZ commented on cpprevolve/revolve/gazebo/brains/Evaluator.cpp:
```diff
@@ -35,36 +38,50 @@ double Evaluator::measure_distance(
 
 /////////////////////////////////////////////////
-void Evaluator::Reset()
+void Evaluator::reset()
```
No more camel case for functions?

DaanZ commented on cpprevolve/revolve/gazebo/plugin/RobotController.cpp:
```diff
@@ -199,44 +203,70 @@ SensorFactoryPtr RobotController::SensorFactory(
-  std::cout << "Loading controller " << controller_type << " and learner " << learner << std::endl;
+    auto brain_sdf = _sdf->GetElement("rv:brain");
+    auto controller_type = brain_sdf->GetElement("rv:controller")->GetAttribute("type")->GetAsString();
+    auto learner_type = brain_sdf->GetElement("rv:learner")->GetAttribute("type")->GetAsString();
```
Revert tabs?

DaanZ commented on cpprevolve/revolve/gazebo/plugin/RobotController.cpp:
```diff
@@ -265,10 +295,28 @@ void RobotController::CheckUpdate(const ::gazebo::common::UpdateInfo _info)
 {
-  auto currentTime = _info.simTime.Double() - initTime_;
+    const gz::common::Time current_time = _info.simTime - initTime_;
+    const double current_time_d = current_time.Double();
```
Revert tabs?

DaanZ commented on cpprevolve/revolve/gazebo/plugin/RobotController.cpp:
```diff
@@ -265,10 +295,28 @@ void RobotController::CheckUpdate(const ::gazebo::common::UpdateInfo _info)
+        evaluator->simulation_update(world_pose, current_time_d, delta_time);
+    }
 
-  if (brain_)
```
Where did the brain go?

DaanZ reviewed (commented):
I am a bit worried that some of the changes will affect the way we do experiments in the future, and I am not sure how much of these changes will actually be reusable.

Commit:
General:
 - Continue run
 - Load best genome

IMC:
 - Modularity from yaml
 - open-loop logging

BO:
 - LHS initialization

Evaluator:
 - Logging components of the fitnesses
 - Reset robot correctly with fitness

Positionmotor:
 - Added possibility for motor input noise

python_code:
 - Added yaml functionality for IMC

plane.world:
 - Added realtime and camera #commented <!-- -->

Commit:
General:
 - Continue run
 - Load best genome

IMC:
 - Modularity from yaml
 - open-loop logging

BO:
 - LHS initialization

Evaluator:
 - Logging components of the fitnesses
 - Reset robot correctly with fitness

Positionmotor:
 - Added possibility for motor input noise

python_code:
 - Added yaml functionality for IMC

plane.world:
 - Added realtime and camera #commented <!-- -->

Commit:
General:
 - Continue run
 - Load best genome

IMC:
 - Modularity from yaml
 - open-loop logging

BO:
 - LHS initialization

Evaluator:
 - Logging components of the fitnesses
 - Reset robot correctly with fitness

Positionmotor:
 - Added possibility for motor input noise

python_code:
 - Added yaml functionality for IMC

plane.world:
 - Added realtime and camera #commented <!-- -->

Commit:
General:
 - Continue run
 - Load best genome

IMC:
 - Modularity from yaml
 - open-loop logging

BO:
 - LHS initialization

Evaluator:
 - Logging components of the fitnesses
 - Reset robot correctly with fitness

Positionmotor:
 - Added possibility for motor input noise

python_code:
 - Added yaml functionality for IMC

plane.world:
 - Added realtime and camera #commented <!-- -->

DaanZ committed:
Merge GECCO paper branch from Fuda van Diggelen
